lib/elixir/lib/enum.ex

```diff
@@ -717,30 +717,64 @@ defmodule Enum do
   """
   @spec fetch(t, integer) :: {:ok, element} | :error
   def fetch(enumerable, index)
+  def fetch([], index) when is_integer(index),
+    do: :error
+  def fetch([elem], 0),
+    do: {:ok, elem}
+  def fetch([elem], -1),
+    do: {:ok, elem}
+  def fetch([_], index) when is_integer(index),
+    do: :error
+
+  def fetch(enumerable, index)
+      when is_map(enumerable) and map_size(enumerable) == 0 and is_integer(index),
+    do: :error
+  def fetch(enumerable, 0) when is_map(enumerable) and map_size(enumerable) == 1,
+    do: :error
+  def fetch(enumerable, -1) when is_map(enumerable) and map_size(enumerable) == 1,
+    do: :error
+  def fetch(enumerable, index)
+      when is_map(enumerable) and is_integer(index) and map_size(enumerable) == 1,
+    do: :error
+
+  def fetch(enumerable, index) when is_integer(index) and index < 0 do
+    case count(enumerable) + index do
```

Comment:
I was not commenting on the performance, I was commenting on the correctness. :) Those functions should traverse the enumerable just once. :)